Settings generator: use single Mac agent for deployment

ilya-g opened the pull request.

ilya-g commented:
I've decided to rework the approach a bit: instead of doing it unconditionally, introduce an option for that, because there can be situations where cross-compilation on macs may be insufficient.

ilya-g commented on main/src/kotlinx/team/infra/Publishing.kt:
```diff
@@ -33,6 +33,11 @@ open class PublishingConfiguration @Inject constructor(val objects: ObjectFactor
+    /**
+     * Build all artifacts for publishing and deploy them from a single mac agent.
+     */
+    var singleAgentMacDeployment: Boolean = false
```
Perhaps, it'd be better to place it in `teamcity` DSL block because it affects generated teamcity build configurations?
On the other hand, `libraryStagingRepoDescription` was recently moved (see the commit 25e365088a) from `teamcity` to `publishing/sonatype` block, because while it also affects TC configurations, it is semantically closer to sonatype publishing.
WDYT?

qurbonzoda replied:
I think "publishing" DSL is the appropriate place for the property. It affects publishing, even if only in TeamCity.

qurbonzoda commented:
Is the ability to publish artifacts for all targets documented somewhere?
I just need to better understand the cross-compilation feature in order to review the MR properly.

ilya-g commented:
Cross-compiling for Linux on Mac hosts existed for a long time, and cross-compiling for Windows was announced in one of recent releases: https://kotlinlang.org/docs/whatsnew16.html#compilation-of-windows-targets-on-any-host

But in my experiment, the artifacts I've got on a Mac agent were slightly different, so I'd like to study it a bit more before enabling for production publishing.

fzhinkin committed:
Settings generator: introduce option to use single Mac agent for deployment

# Conflicts:
#	main/resources/teamcity/settings.kts
#	main/src/kotlinx/team/infra/TeamCity.kt

# Conflicts:
#	main/resources/teamcity/utils.kt

fzhinkin commented:
^ Rebased the branch onto master

fzhinkin commented on main/resources/teamcity/settings.kts:
```diff
@@ -56,7 +56,7 @@ project {
         val deployVersion = deployVersion()
         val deployAll = deployAll(deployVersion)
-        val deploys = platforms.map { buildArtifacts(deployVersion, it) }
+        val deploys = platforms.associateWith { buildArtifacts(deployVersion, it) }
```
@ilya-g, why build tasks has to be created for all platforms even if the `singleAgentMacDeployment` is `true`? Does it make sense to skip their creation in the first place, instead of omitting them from publication task's dependencies?

ilya-g replied:
Since we're mostly switching the existing projects with the established deployment builds, I believe the idea was to preserve these builds so that their history is not lost and so that we could switch back after some time if something goes wrong.

ilya-g replied:
We can always remove the unnecessary build types (by not creating them) later if we deem this new scheme satisfactory.

ilya-g commented on main/src/kotlinx/team/infra/Publishing.kt:
```diff
@@ -21,6 +21,11 @@ open class PublishingConfiguration @Inject constructor() {
+    /**
+     * Build all artifacts for publishing and deploy them from a single mac agent.
+     */
+    var singleAgentMacDeployment: Boolean = false
```
Perhaps we can now switch it to `true` by default

fzhinkin replied:
Makes sense. Flipped the value (and rebased the branch once again)

fzhinkin committed:
Flipped singleAgentMacDeployment default value

ilya-g commented:
All good 👍 
Can't approve because I'm an author of this PR 😄
